### ISSUE TITLE
weighted approach to multinormal models with residual dependency in brms

### ISSUE THREAD

**Issue by skoval** [paul-buerkner/brms#910]
This issue picks up from the following thread on [stan discourse](https://discourse.mc-stan.org/t/why-is-combining-response-weights-with-ar-cov-true-not-possible-in-brms/15277).

This issue concerns normal models with residual dependency. Currently, it is not possible to have weights as there is not a pointwise implementation of the likelihood. An example of the kind of model for which this occurs are some class of normal models with autocorrelation, such as

`
y | weights(w) ~ ar(cov = TRUE) + ...
`

Models of this type would currently invoke `normal_time_hom_lpdf`.

Thanks to the results from the work on pointwise likelihoods (https://arxiv.org/abs/1810.10559), I have tested a pointwise version of the equivalent likelihood. The function `normal_pointwise_time_hom_lpdf` is the observation-level likelihood with as few other alterations from `normal_time_hom_lpdf` as I could make. This involves sequential application of the leave-one-out result, so there is still a matrix inversion with each iteration although the dimensions of the matrix decreases with each step. Similar results for a student T are included in  https://arxiv.org/abs/1810.10559 so that should be a straightforward extension.

I'm guessing there is a more efficient and elegant way to do this and hope that this initial start makes it easier for other developers to see what that might be. 

```
real loo_normal_cholesky(vector y, vector mu, matrix Sigma){
    int n = dims(y)[1];
    matrix[n, n] C0;
    matrix[n, n] C;
    vector[n] R;
    vector[n] G;
    real cbar;
    real g1;
    real mu1;
    real sigma1;

      C0  = inverse(Sigma);
      C = C0' * C0;
      R = y - mu;
      cbar = C[1,1];
      G = C * R;
      g1 = G[1];
      mu1 = y[1] - g1/cbar;
      sigma1 = sqrt(1/cbar);
    
  return normal_lpdf(y[1] | mu1, sigma1);
  }  
```

```
real normal_pointwise_time_hom_lpdf(vector y, vector mu, real sigma, matrix chol_cor, 
                            vector se2, int[] nobs, int[] begin, int[] end, vector w) {
    int I = size(nobs);
    int has_se = max(se2) > 0;
    vector[I] lp; 

    for (i in 1:I) { 
      matrix[nobs[i], nobs[i]] L;
      vector[nobs[i]] y0;
      vector[nobs[i]] mu0;
      vector[nobs[i]] lp0;
      
      L = sigma * chol_cor[1:nobs[i], 1:nobs[i]];
      if (has_se) {
        // need to add 'se' to the correlation matrix itself
        L = multiply_lower_tri_self_transpose(L);
        L += diag_matrix(se2[begin[i]:end[i]]);
        L = cholesky_decompose(L);
      }

      y0 = y[begin[i]:end[i]];
      mu0 = mu[begin[i]:end[i]];

      for(j in 1:nobs[i])
          lp0[j] = loo_normal_cholesky(y0[j:nobs[i]], mu0[j:nobs[i]], L[j:nobs[i],j:nobs[i]]);

     lp[i] = dot_product(w[begin[i]:end[i]], lp0);
    }                        
    return sum(lp); 
  }
```

**Comment by paul-buerkner** [comment]
Thank you. This is a really good start! We should get away with one matrix inversion per block (instead of one per step with decreasing dimension) and I will update the code accordingly when implementing this issue.

**Comment by martiningram** [comment]
Hi there, I was wondering if this could be handled using the fact that we can transform the MVN into a standard normal variable by multiplying with the inverse of the Cholesky? I.e. (in R):

```
weighted_log_lik <- function(x, weights, mean, covariance) {
  
  # Variable transformation
  L <- t(chol(covariance))
  z <- solve(L, x - mu)
  
  # Log Jacobian correction
  L_logdet <- determinant(L, logarithm = TRUE)
  L_logdet <- L_logdet$sign * L_logdet$modulus
  
  # Weighted sum of the result, adjusted by the Jacobian
  z_dens <- sum(weights * dnorm(z, log = TRUE)) - L_logdet
  
  as.numeric(z_dens)
  
}

weighted_log_lik(vec, rep(1, n), mu, cov)
```

However, I don't know whether this is valid, since we will be multiplying the transformed likelihoods by the weights, which may not be the same as weighting the individual likelihoods. But I thought I'd mention it in case this is helpful.

**Comment by paul-buerkner** [comment]
I don't know either if this is valid straight away but since we have a proven solution in https://arxiv.org/abs/1810.10559 the correctness should be easy to verify.

**Comment by martiningram** [comment]
Thanks @paul-buerkner I gave it a go for the 2x2 case and it looks like it may not work out to be the same as weighting the sequential conditionals. My apologies, I should have checked this before commenting.

**Comment by paul-buerkner** [comment]
I thought of the problem a little more and I don't think that the efficient LOO densities we found in https://arxiv.org/abs/1810.10559 will help us much. The problem ist that we have to obtain the joint density from individual terms, which works using a standard factorization of the form p(y1) p(y2 | y1) p(y3 | y1, y2) ... 

However, this factorization requires (a) a natural ordering of the ys, which we have in time-series but not in spatial data and (b) and efficient way to compute p(y2 | y1), p(y3 | y1, y2), etc. which we do not have in general at this point because we need to compute matrix inverse (or equivalent) at every step. 

In other words, I don't see this feature being implemented anytime soon, or even at all given that I don't see how this can be mathematically expressed in a general-enough way.

**Comment by skoval** [comment]
Thanks for giving this one more consideration. I agree that while this
might be doable for some class of models, the matrix inversion would likely
limit its practical usefulness without faster inversion methods. One for
the backburner for now.

**Comment by paul-buerkner** [comment]
I will close this issue now as I really don't see a way to implement this properly. If you come up with a new idea, please feel free to reopen this issue.
